### ISSUE TITLE
LoRA fine-tuning: does cell type subset selection affect TFBS specificity predictions?

### ISSUE THREAD

**Issue by cjiang310437** [buenrostrolab/scPrinter#104]
Hi,

I have been using scPrinter for cell-type-specific TFBS prediction using LoRA fine-tuning, and I observed that the choice of cell types included in fine-tuning significantly affects the results.

**Observation:**
When fine-tuning with only the target cell types of interest, the model correctly identifies cell-type-specific enhancer activity consistent with known biology.

When fine-tuning with all available cell types, some cell-type-specific signals are incorrectly attributed to the wrong cell type. We hypothesize this is because cell types with stronger/broader chromatin accessibility dominate the training loss, compressing the embedding space of the target cell types.

**Questions:**
1. Is there a recommended strategy for selecting which cell types to include in LoRA fine-tuning?
2. Does the framework assume any specific properties of the embedding space (e.g., balanced signal strength across cell types)?
3. Would loss re-weighting per cell type be a reasonable modification to reduce this bias?

Thank you!

**Comment by ruochiz** [comment]
Thank you for bringing this up, and I have seen similar effects like this before, and is part of the reason in tutorial here we did a 2-step lora finetune: step 1 finetune on all cell types, step2 further lora finetune on a trajectory of interest.

https://ruochiz.com/scprinter_doc/tutorials/scprinter_seq_BMMC_tutorial_v1.html#Using-the-cell-embeddings-for-LoRA-fintune-seq2PRINT-models

I tried loss re-weight through a dynamic reweighting by gradient norm (so each cell type contribute to the gradient roughly equally) and it helps a little bit. 

Overall the multi-stage lora finetune would be the most straightforward way to achieve similar effects without much changing to the code and takes similar computational cost as compared to the reweighting. 

Hope this information is helpful.

**Comment by cjiang310437** [comment]
> Thank you for bringing this up, and I have seen similar effects like this before, and is part of the reason in tutorial here we did a 2-step lora finetune: step 1 finetune on all cell types, step2 further lora finetune on a trajectory of interest.
> 
> https://ruochiz.com/scprinter_doc/tutorials/scprinter_seq_BMMC_tutorial_v1.html#Using-the-cell-embeddings-for-LoRA-fintune-seq2PRINT-models
> 
> I tried loss re-weight through a dynamic reweighting by gradient norm (so each cell type contribute to the gradient roughly equally) and it helps a little bit.
> 
> Overall the multi-stage lora finetune would be the most straightforward way to achieve similar effects without much changing to the code and takes similar computational cost as compared to the reweighting.
> 
> Hope this information is helpful.

Thank you! I will have a try and provide a feedback!

**Comment by cjiang310437** [comment]
Following up on that, I have a question regarding data balancing across cell types during the fine-tuning stage. In my dataset, some cell types have enough cells to generate 10 pseudobulks, while rarer populations can only produce 3.

1. Does having a significantly higher number of pseudobulks for major cell types exacerbate the bias where they dominate the training loss, even with the two-step approach?

2. Would you recommend downsampling the major cell types to a 'lowest common denominator' (e.g., using 3 pseudobulks for all) to ensure an even signal? Or is it better to keep all available data to provide the model with more variance for those major types?

**Comment by cjiang310437** [comment]
@ruochiz

**Comment by ruochiz** [comment]
So. in the past, we keep all of them, as our pseudobulks are meta-cell guided not just based on cell type label, to capture continuous states of developmental trajectories or just different cell states within the same cell type. 

However, I do see the point of downsampling some of the major celltypes so the information won't be too biased towards them. I would suggest try both if possible.

**Comment by cjiang310437** [comment]
> So. in the past, we keep all of them, as our pseudobulks are meta-cell guided not just based on cell type label, to capture continuous states of developmental trajectories or just different cell states within the same cell type.
> 
> However, I do see the point of downsampling some of the major celltypes so the information won't be too biased towards them. I would suggest try both if possible.

Thank you. Will have a try.
